docs/dev/specs/packable.md

```diff
@@ -0,0 +1,315 @@
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  deserialization.
+
+- The `unpack_bytes` method writes enough bytes to fill `slice`. This method
+  must fail if the unpacker does not have enough bytes to fulfill the request.
+
+We also have the `Packer` trait which represents a type that can be used to
+write bytes into it.
+
+```rust
+pub trait Packer {
+    type Error;
+
+    fn pack_bytes(&mut self, bytes: &[u8]) -> Result<(), Self::Error>;
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  serialization.
+
+- The `pack_bytes` method reads all the bytes from `slice`. This method must
+  fail if the packer does not have enough space to fulfill the request.
+
+Both traits allows us to abstract away any IO operation without relying on
+`std::io`. This has the additional benefit of allowing us to pack and unpack
+from different kinds of buffers without making specific assumptions about the
+implementation of the serialization itself.
+
+## The `Packable` trait
+
+The new `Packable` trait is the following
+
+```rust
+pub trait Packable: Sized {
+    type Error;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
+
+    fn packed_len(&self) -> usize;
+}
+```
+
+- The `Error` associated type represents a semantic error.
+
+- The `pack` method serializes the current value using a `Packer` to write the
+  bytes.
+
+- The `unpack` method deserializes a value using an `Unpacker` to read the
+  bytes.
+
+- The `packed_len` method returns the length in bytes of the serialized value.
+  This has to match the number of bytes written using `pack` to avoid
+  inconsistencies.
+
+## Error handling
+
+In addition to the three `Error` associated types mentioned before, we provide
+two new helper error types:
+
+- The `UnpackError` type which is an enum wrapping values of either the
+  `Unpacker::Error` or `Packable::Error` type.
+
+- The `UnknownTagError` type which can be used if the prefix tag used to
+  represent the variant of an enum does not correspond to any variant of such
+  enum. As a convention, any `Packable::Error` type for an enum must implement
+  `From<UnknownTagError>` (this is not enforced in the trait itself because
+  this error is not used for structs).
+
+We also use `core::convert::Infallible` as `Packable::Error` for types whose
+unpacking does not have semantic errors, like integers for example. A more
+grounded explanation of error handling can be found in the *the `derive` macro*
+section.
+
+## `Packable` for basic types
+
+The `Packable` is implemented for every integer type by encoding the value as
+an array of bytes in little-endian order. Booleans are packed following Rust's
+data layout, meaning that `true` is packed as a `1` byte and `false` as a `0`
+byte. However, boolean unpacking is less strict and unpacks any non-zero byte
+as `true`.
+
```

Comment:
Should we mention we also provide implementations for arrays, vectors and arrays when the type is itself packable ?